server/app/services/program/ProgramService.java

```diff
@@ -583,6 +587,38 @@ public ErrorAnd<ProgramDefinition, CiviFormError> updateProgramDefinition(
             .join());
   }
 
+  /** Preserve translations on existing titles and descriptions in application steps */
+  private ImmutableList<ApplicationStep> preserveApplicationStepTranslations(
+      ImmutableList<ApplicationStep> newSteps, ImmutableList<ApplicationStep> existingSteps) {
+    ImmutableMap<String, LocalizedStrings> existingApplicationStepsTitlesMap =
+        existingSteps.stream()
+            .collect(
+                ImmutableMap.toImmutableMap(
+                    step -> step.getTitle().getDefault(), step -> step.getTitle()));
+    ImmutableMap<String, LocalizedStrings> existingApplicationStepsDescriptionsMap =
+        existingSteps.stream()
+            .collect(
+                ImmutableMap.toImmutableMap(
+                    step -> step.getDescription().getDefault(), step -> step.getDescription()));
+
+    return newSteps.stream()
+        .map(
+            step -> {
+              String defaultTitle = step.getTitle().getDefault();
+              String defaultDescription = step.getDescription().getDefault();
+              if (existingApplicationStepsTitlesMap.containsKey(defaultTitle)) {
+                step = step.setTitle(existingApplicationStepsTitlesMap.get(defaultTitle));
+              }
+              if (existingApplicationStepsDescriptionsMap.containsKey(defaultDescription)) {
+                step =
+                    step.setDescription(
+                        existingApplicationStepsDescriptionsMap.get(defaultDescription));
+              }
+              return step;
+            })
+        .collect(ImmutableList.toImmutableList());
```

Comment:
I like it! I think it's easier to read. I was trying to find a way to avoid doing a loop within a loop, although maybe a stream within a stream is more performant? Also at this point we'll only have at most 5 steps, so it's probably not so bad.